string_decode.lua

```diff
@@ -2,42 +2,21 @@
 -- String decoder functions
 --
 
-local stringdecode = {}
-
--- From http://lua-users.org/wiki/LuaUnicode
-local function utf8_to_32(utf8str)
-    assert(type(utf8str) == "string")
-    local res, seq, val = {}, 0, nil
-
-    for i = 1, #utf8str do
-        local c = string.byte(utf8str, i)
-        if seq == 0 then
-            table.insert(res, val)
-            seq = c < 0x80 and 1 or c < 0xE0 and 2 or c < 0xF0 and 3 or
-                  c < 0xF8 and 4 or --c < 0xFC and 5 or c < 0xFE and 6 or
-                error("Invalid UTF-8 character sequence")
-            val = bit32.band(c, 2^(8-seq) - 1)
-        else
-            val = bit32.bor(bit32.lshift(val, 6), bit32.band(c, 0x3F))
-        end
-
-        seq = seq - 1
-    end
+local lunicode = require(".unicode")
 
-    table.insert(res, val)
-
-    return res
-end
+local stringdecode = {}
 
 function stringdecode.decode(str, encoding)
     local enc = encoding and encoding:lower() or "ascii"
 
     if enc == "ascii" then
         return str
     elseif enc == "utf-8" then
-        local code_points = utf8_to_32(str)
-
-        return utf8.char(table.unpack(code_points))
+	return lunicode.transcode(str, lunicode.utf8_dec, lunicode.utf8_enc, false, false)
+    elseif enc == "sjis" then
+	return table.concat(lunicode.decode(str, lunicode.sjis_dec, true))
+    elseif enc == "cp437" then
+        return lunicode.transcode(str, lunicode.cp437_dec, lunicode.utf8_enc, false, false)
```

Comment:
Is there really no fully fledged encoding library in Lua (some port /wrapper of `iconv` lib for example) that could handle more than these 4 encodings?

Reply:
There of course is, but I tried hard not to introduce more external dependencies, and specifically not dependencies that bind to platform-native compiled binaries. It is in our best interest imho to keep the lua runtime pure-lua, for best cross-platform/cross-environment compatibility